// langsmith/dist/_openapi_client/resources/public/datasets.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Datasets = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const pagination_js_1 = require("../../core/pagination.cjs");
const headers_js_1 = require("../../internal/headers.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Datasets extends resource_js_1.APIResource {
    /**
     * Get dataset by ids or the shared dataset if not specifed.
     */
    list(shareToken, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/public/${shareToken}/datasets`, { query, ...options });
    }
    /**
     * Get all comparative experiments for a given dataset.
     */
    listComparative(shareToken, query = {}, options) {
        return this._client.getAPIList((0, path_js_1.path) `/api/v1/public/${shareToken}/datasets/comparative`, (pagination_js_1.OffsetPaginationTopLevelArray), { query, ...options });
    }
    /**
     * Get feedback for runs in projects run over a dataset that has been shared.
     */
    listFeedback(shareToken, query = {}, options) {
        return this._client.getAPIList((0, path_js_1.path) `/api/v1/public/${shareToken}/datasets/feedback`, (pagination_js_1.OffsetPaginationTopLevelArray), { query, ...options });
    }
    /**
     * Get projects run on a dataset that has been shared.
     */
    listSessions(shareToken, params = {}, options) {
        const { accept, ...query } = params ?? {};
        return this._client.getAPIList((0, path_js_1.path) `/api/v1/public/${shareToken}/datasets/sessions`, (pagination_js_1.OffsetPaginationTopLevelArray), {
            query,
            ...options,
            headers: (0, headers_js_1.buildHeaders)([{ ...(accept != null ? { accept: accept } : undefined) }, options?.headers]),
        });
    }
    /**
     * Get sessions from multiple datasets using share tokens.
     */
    retrieveSessionsBulk(query, options) {
        return this._client.get('/api/v1/public/datasets/sessions-bulk', { query, ...options });
    }
}
exports.Datasets = Datasets;
