// langsmith/dist/_openapi_client/resources/public/datasets.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import { OffsetPaginationTopLevelArray, } from '../../core/pagination.js';
import { buildHeaders } from '../../internal/headers.js';
import { path } from '../../internal/utils/path.js';
export class Datasets extends APIResource {
    /**
     * Get dataset by ids or the shared dataset if not specifed.
     */
    list(shareToken, query = {}, options) {
        return this._client.get(path `/api/v1/public/${shareToken}/datasets`, { query, ...options });
    }
    /**
     * Get all comparative experiments for a given dataset.
     */
    listComparative(shareToken, query = {}, options) {
        return this._client.getAPIList(path `/api/v1/public/${shareToken}/datasets/comparative`, (OffsetPaginationTopLevelArray), { query, ...options });
    }
    /**
     * Get feedback for runs in projects run over a dataset that has been shared.
     */
    listFeedback(shareToken, query = {}, options) {
        return this._client.getAPIList(path `/api/v1/public/${shareToken}/datasets/feedback`, (OffsetPaginationTopLevelArray), { query, ...options });
    }
    /**
     * Get projects run on a dataset that has been shared.
     */
    listSessions(shareToken, params = {}, options) {
        const { accept, ...query } = params ?? {};
        return this._client.getAPIList(path `/api/v1/public/${shareToken}/datasets/sessions`, (OffsetPaginationTopLevelArray), {
            query,
            ...options,
            headers: buildHeaders([{ ...(accept != null ? { accept: accept } : undefined) }, options?.headers]),
        });
    }
    /**
     * Get sessions from multiple datasets using share tokens.
     */
    retrieveSessionsBulk(query, options) {
        return this._client.get('/api/v1/public/datasets/sessions-bulk', { query, ...options });
    }
}
