// langsmith/dist/_openapi_client/resources/feedback/feedback.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import * as ConfigsAPI from './configs.js';
import { Configs } from './configs.js';
import * as TokensAPI from './tokens.js';
import { Tokens, } from './tokens.js';
import { OffsetPaginationTopLevelArray, } from '../../core/pagination.js';
import { path } from '../../internal/utils/path.js';
export class Feedback extends APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "tokens", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new TokensAPI.Tokens(this._client)
        });
        Object.defineProperty(this, "configs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ConfigsAPI.Configs(this._client)
        });
    }
    /**
     * Create a new feedback.
     */
    create(body, options) {
        return this._client.post('/api/v1/feedback', { body, ...options });
    }
    /**
     * Get a specific feedback.
     */
    retrieve(feedbackID, query = {}, options) {
        return this._client.get(path `/api/v1/feedback/${feedbackID}`, { query, ...options });
    }
    /**
     * Replace an existing feedback entry with a new, modified entry.
     */
    update(feedbackID, body, options) {
        return this._client.patch(path `/api/v1/feedback/${feedbackID}`, { body, ...options });
    }
    /**
     * List all Feedback by query params.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/v1/feedback', (OffsetPaginationTopLevelArray), {
            query,
            ...options,
        });
    }
    /**
     * Delete a feedback.
     */
    delete(feedbackID, options) {
        return this._client.delete(path `/api/v1/feedback/${feedbackID}`, options);
    }
}
Feedback.Tokens = Tokens;
Feedback.Configs = Configs;
