// langsmith/dist/_openapi_client/resources/feedback/feedback.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Feedback = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const ConfigsAPI = __importStar(require("./configs.cjs"));
const configs_js_1 = require("./configs.cjs");
const TokensAPI = __importStar(require("./tokens.cjs"));
const tokens_js_1 = require("./tokens.cjs");
const pagination_js_1 = require("../../core/pagination.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Feedback extends resource_js_1.APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "tokens", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new TokensAPI.Tokens(this._client)
        });
        Object.defineProperty(this, "configs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ConfigsAPI.Configs(this._client)
        });
    }
    /**
     * Create a new feedback.
     */
    create(body, options) {
        return this._client.post('/api/v1/feedback', { body, ...options });
    }
    /**
     * Get a specific feedback.
     */
    retrieve(feedbackID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/feedback/${feedbackID}`, { query, ...options });
    }
    /**
     * Replace an existing feedback entry with a new, modified entry.
     */
    update(feedbackID, body, options) {
        return this._client.patch((0, path_js_1.path) `/api/v1/feedback/${feedbackID}`, { body, ...options });
    }
    /**
     * List all Feedback by query params.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/v1/feedback', (pagination_js_1.OffsetPaginationTopLevelArray), {
            query,
            ...options,
        });
    }
    /**
     * Delete a feedback.
     */
    delete(feedbackID, options) {
        return this._client.delete((0, path_js_1.path) `/api/v1/feedback/${feedbackID}`, options);
    }
}
exports.Feedback = Feedback;
Feedback.Tokens = tokens_js_1.Tokens;
Feedback.Configs = configs_js_1.Configs;
