// @langchain/langgraph-checkpoint/dist/base.cjs
const require_id = require("./id.cjs");
const require_types = require("./serde/types.cjs");
const require_jsonplus = require("./serde/jsonplus.cjs");
//#region src/base.ts
function deepCopy(obj) {
	if (typeof obj !== "object" || obj === null) return obj;
	const newObj = Array.isArray(obj) ? [] : {};
	for (const key in obj) if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = deepCopy(obj[key]);
	return newObj;
}
/** @hidden */
function emptyCheckpoint() {
	return {
		v: 4,
		id: require_id.uuid6(0),
		ts: (/* @__PURE__ */ new Date()).toISOString(),
		channel_values: {},
		channel_versions: {},
		versions_seen: {}
	};
}
/** @hidden */
function copyCheckpoint(checkpoint) {
	return {
		v: checkpoint.v,
		id: checkpoint.id,
		ts: checkpoint.ts,
		channel_values: { ...checkpoint.channel_values },
		channel_versions: { ...checkpoint.channel_versions },
		versions_seen: deepCopy(checkpoint.versions_seen)
	};
}
var BaseCheckpointSaver = class {
	serde = new require_jsonplus.JsonPlusSerializer();
	constructor(serde) {
		this.serde = serde || this.serde;
	}
	async get(config) {
		const value = await this.getTuple(config);
		return value ? value.checkpoint : void 0;
	}
	/**
	* Walk the parent chain returning per-channel writes + seed, used to
	* reconstruct `DeltaChannel` state from `checkpoint_writes`.
	*
	* For each requested channel, walks ancestors of the checkpoint identified
	* by `config` (following `parentConfig`) and accumulates the pending writes
	* for that channel. The walk terminates per-channel at the nearest ancestor
	* whose `channel_values[ch]` is populated; that value is returned as `seed`.
	* If the walk reaches the root without finding a stored value, `seed` is
	* omitted from that channel's entry — the consumer treats the absence as
	* "start empty".
	*
	* Walks the parent chain (not `list({ before })`): for forked threads, only
	* on-path ancestors contribute.
	*
	* The default implementation walks `getTuple` + `parentConfig` once for all
	* channels — each ancestor visited once, not once per channel. Savers with
	* direct storage access (e.g. `MemorySaver`) override for performance; the
	* return contract is fixed here.
	*
	* @remarks Beta. The signature, return shape, and interaction with
	* `DeltaSnapshot` blobs may change. Override at your own risk; the default
	* implementation will continue to work against the public
	* `BaseCheckpointSaver` contract.
	*
	* @param options.config Configuration identifying the target checkpoint.
	* @param options.channels Channel names to walk for. Empty → empty mapping.
	* @returns Per-channel {@link DeltaChannelHistory} for every requested name.
	*/
	async getDeltaChannelHistory(options) {
		const { config, channels } = options;
		if (channels.length === 0) return {};
		const collectedByCh = {};
		const seedByCh = {};
		const remaining = new Set(channels);
		for (const ch of channels) collectedByCh[ch] = [];
		let cursorConfig = (await this.getTuple(config))?.parentConfig;
		while (cursorConfig != null && remaining.size > 0) {
			const tup = await this.getTuple(cursorConfig);
			if (tup === void 0) break;
			if (tup.pendingWrites && tup.pendingWrites.length > 0) for (let i = tup.pendingWrites.length - 1; i >= 0; i -= 1) {
				const write = tup.pendingWrites[i];
				const ch = write[1];
				if (remaining.has(ch)) collectedByCh[ch].push(write);
			}
			for (const ch of Array.from(remaining)) if (Object.prototype.hasOwnProperty.call(tup.checkpoint.channel_values, ch)) {
				seedByCh[ch] = tup.checkpoint.channel_values[ch];
				remaining.delete(ch);
			}
			cursorConfig = tup.parentConfig;
		}
		const result = {};
		for (const ch of channels) {
			const entry = { writes: collectedByCh[ch].slice().reverse() };
			if (Object.prototype.hasOwnProperty.call(seedByCh, ch)) entry.seed = seedByCh[ch];
			result[ch] = entry;
		}
		return result;
	}
	/**
	* Generate the next version ID for a channel.
	*
	* Default is to use integer versions, incrementing by 1. If you override, you can use str/int/float versions,
	* as long as they are monotonically increasing.
	*/
	getNextVersion(current) {
		if (typeof current === "string") throw new Error("Please override this method to use string versions.");
		return current !== void 0 && typeof current === "number" ? current + 1 : 1;
	}
};
function compareChannelVersions(a, b) {
	if (typeof a === "number" && typeof b === "number") return Math.sign(a - b);
	return String(a).localeCompare(String(b));
}
function maxChannelVersion(...versions) {
	return versions.reduce((max, version, idx) => {
		if (idx === 0) return version;
		return compareChannelVersions(max, version) >= 0 ? max : version;
	});
}
/**
* Mapping from error type to error index.
* Regular writes just map to their index in the list of writes being saved.
* Special writes (e.g. errors) map to negative indices, to avoid those writes from
* conflicting with regular writes.
* Each Checkpointer implementation should use this mapping in put_writes.
*/
const WRITES_IDX_MAP = {
	[require_types.ERROR]: -1,
	[require_types.SCHEDULED]: -2,
	[require_types.INTERRUPT]: -3,
	[require_types.RESUME]: -4
};
function getCheckpointId(config) {
	return config.configurable?.checkpoint_id || config.configurable?.thread_ts || "";
}
//#endregion
exports.BaseCheckpointSaver = BaseCheckpointSaver;
exports.WRITES_IDX_MAP = WRITES_IDX_MAP;
exports.compareChannelVersions = compareChannelVersions;
exports.copyCheckpoint = copyCheckpoint;
exports.deepCopy = deepCopy;
exports.emptyCheckpoint = emptyCheckpoint;
exports.getCheckpointId = getCheckpointId;
exports.maxChannelVersion = maxChannelVersion;
