// langsmith/dist/_openapi_client/resources/annotation-queues/annotation-queues.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnotationQueues = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const RunsAPI = __importStar(require("./runs.cjs"));
const runs_js_1 = require("./runs.cjs");
const pagination_js_1 = require("../../core/pagination.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class AnnotationQueues extends resource_js_1.APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "runs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RunsAPI.Runs(this._client)
        });
    }
    /**
     * Get Annotation Queue
     */
    retrieve(queueID, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}`, options);
    }
    /**
     * Update Annotation Queue
     */
    update(queueID, body, options) {
        return this._client.patch((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}`, { body, ...options });
    }
    /**
     * Delete Annotation Queue
     */
    delete(queueID, options) {
        return this._client.delete((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}`, options);
    }
    /**
     * Create Annotation Queue
     */
    annotationQueues(body, options) {
        return this._client.post('/api/v1/annotation-queues', { body, ...options });
    }
    /**
     * Create Identity Annotation Queue Run Status
     */
    createRunStatus(annotationQueueRunID, body, options) {
        return this._client.post((0, path_js_1.path) `/api/v1/annotation-queues/status/${annotationQueueRunID}`, {
            body,
            ...options,
        });
    }
    /**
     * Export Annotation Queue Archived Runs
     */
    export(queueID, body, options) {
        return this._client.post((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/export`, { body, ...options });
    }
    /**
     * Populate annotation queue with runs from an experiment.
     */
    populate(body, options) {
        return this._client.post('/api/v1/annotation-queues/populate', { body, ...options });
    }
    /**
     * Get Annotation Queues
     */
    retrieveAnnotationQueues(query = {}, options) {
        return this._client.getAPIList('/api/v1/annotation-queues', (pagination_js_1.OffsetPaginationTopLevelArray), { query, ...options });
    }
    /**
     * Get Annotation Queues For Run
     */
    retrieveQueues(runID, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${runID}/queues`, options);
    }
    /**
     * Get a run from an annotation queue
     */
    retrieveRun(index, params, options) {
        const { queue_id, ...query } = params;
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${queue_id}/run/${index}`, { query, ...options });
    }
    /**
     * Get Size From Annotation Queue
     */
    retrieveSize(queueID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/size`, { query, ...options });
    }
    /**
     * Get Total Archived From Annotation Queue
     */
    retrieveTotalArchived(queueID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/total_archived`, { query, ...options });
    }
    /**
     * Get Total Size From Annotation Queue
     */
    retrieveTotalSize(queueID, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/total_size`, options);
    }
}
exports.AnnotationQueues = AnnotationQueues;
AnnotationQueues.Runs = runs_js_1.Runs;
