// langsmith/dist/_openapi_client/resources/annotation-queues/annotation-queues.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import * as RunsAPI from './runs.js';
import { Runs, } from './runs.js';
import { OffsetPaginationTopLevelArray, } from '../../core/pagination.js';
import { path } from '../../internal/utils/path.js';
export class AnnotationQueues extends APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "runs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RunsAPI.Runs(this._client)
        });
    }
    /**
     * Get Annotation Queue
     */
    retrieve(queueID, options) {
        return this._client.get(path `/api/v1/annotation-queues/${queueID}`, options);
    }
    /**
     * Update Annotation Queue
     */
    update(queueID, body, options) {
        return this._client.patch(path `/api/v1/annotation-queues/${queueID}`, { body, ...options });
    }
    /**
     * Delete Annotation Queue
     */
    delete(queueID, options) {
        return this._client.delete(path `/api/v1/annotation-queues/${queueID}`, options);
    }
    /**
     * Create Annotation Queue
     */
    annotationQueues(body, options) {
        return this._client.post('/api/v1/annotation-queues', { body, ...options });
    }
    /**
     * Create Identity Annotation Queue Run Status
     */
    createRunStatus(annotationQueueRunID, body, options) {
        return this._client.post(path `/api/v1/annotation-queues/status/${annotationQueueRunID}`, {
            body,
            ...options,
        });
    }
    /**
     * Export Annotation Queue Archived Runs
     */
    export(queueID, body, options) {
        return this._client.post(path `/api/v1/annotation-queues/${queueID}/export`, { body, ...options });
    }
    /**
     * Populate annotation queue with runs from an experiment.
     */
    populate(body, options) {
        return this._client.post('/api/v1/annotation-queues/populate', { body, ...options });
    }
    /**
     * Get Annotation Queues
     */
    retrieveAnnotationQueues(query = {}, options) {
        return this._client.getAPIList('/api/v1/annotation-queues', (OffsetPaginationTopLevelArray), { query, ...options });
    }
    /**
     * Get Annotation Queues For Run
     */
    retrieveQueues(runID, options) {
        return this._client.get(path `/api/v1/annotation-queues/${runID}/queues`, options);
    }
    /**
     * Get a run from an annotation queue
     */
    retrieveRun(index, params, options) {
        const { queue_id, ...query } = params;
        return this._client.get(path `/api/v1/annotation-queues/${queue_id}/run/${index}`, { query, ...options });
    }
    /**
     * Get Size From Annotation Queue
     */
    retrieveSize(queueID, query = {}, options) {
        return this._client.get(path `/api/v1/annotation-queues/${queueID}/size`, { query, ...options });
    }
    /**
     * Get Total Archived From Annotation Queue
     */
    retrieveTotalArchived(queueID, query = {}, options) {
        return this._client.get(path `/api/v1/annotation-queues/${queueID}/total_archived`, { query, ...options });
    }
    /**
     * Get Total Size From Annotation Queue
     */
    retrieveTotalSize(queueID, options) {
        return this._client.get(path `/api/v1/annotation-queues/${queueID}/total_size`, options);
    }
}
AnnotationQueues.Runs = Runs;
