// langsmith/dist/_openapi_client/resources/annotation-queues/runs.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import { path } from '../../internal/utils/path.js';
export class Runs extends APIResource {
    /**
     * Add Runs To Annotation Queue
     */
    create(queueID, params, options) {
        const { body } = params;
        return this._client.post(path `/api/v1/annotation-queues/${queueID}/runs`, { body: body, ...options });
    }
    /**
     * Update Run In Annotation Queue
     */
    update(queueRunID, params, options) {
        const { queue_id, ...body } = params;
        return this._client.patch(path `/api/v1/annotation-queues/${queue_id}/runs/${queueRunID}`, {
            body,
            ...options,
        });
    }
    /**
     * Get Runs From Annotation Queue
     */
    list(queueID, query = {}, options) {
        return this._client.get(path `/api/v1/annotation-queues/${queueID}/runs`, { query, ...options });
    }
    /**
     * Delete Runs From Annotation Queue
     */
    deleteAll(queueID, body, options) {
        return this._client.post(path `/api/v1/annotation-queues/${queueID}/runs/delete`, { body, ...options });
    }
    /**
     * Delete Run From Annotation Queue
     */
    deleteQueue(queueRunID, params, options) {
        const { queue_id } = params;
        return this._client.delete(path `/api/v1/annotation-queues/${queue_id}/runs/${queueRunID}`, options);
    }
}
