// langsmith/dist/_openapi_client/resources/annotation-queues/runs.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runs = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Runs extends resource_js_1.APIResource {
    /**
     * Add Runs To Annotation Queue
     */
    create(queueID, params, options) {
        const { body } = params;
        return this._client.post((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/runs`, { body: body, ...options });
    }
    /**
     * Update Run In Annotation Queue
     */
    update(queueRunID, params, options) {
        const { queue_id, ...body } = params;
        return this._client.patch((0, path_js_1.path) `/api/v1/annotation-queues/${queue_id}/runs/${queueRunID}`, {
            body,
            ...options,
        });
    }
    /**
     * Get Runs From Annotation Queue
     */
    list(queueID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/runs`, { query, ...options });
    }
    /**
     * Delete Runs From Annotation Queue
     */
    deleteAll(queueID, body, options) {
        return this._client.post((0, path_js_1.path) `/api/v1/annotation-queues/${queueID}/runs/delete`, { body, ...options });
    }
    /**
     * Delete Run From Annotation Queue
     */
    deleteQueue(queueRunID, params, options) {
        const { queue_id } = params;
        return this._client.delete((0, path_js_1.path) `/api/v1/annotation-queues/${queue_id}/runs/${queueRunID}`, options);
    }
}
exports.Runs = Runs;
