// langsmith/dist/_openapi_client/resources/examples/examples.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Examples = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const BulkAPI = __importStar(require("./bulk.cjs"));
const bulk_js_1 = require("./bulk.cjs");
const ValidateAPI = __importStar(require("./validate.cjs"));
const validate_js_1 = require("./validate.cjs");
const pagination_js_1 = require("../../core/pagination.cjs");
const uploads_js_1 = require("../../internal/uploads.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Examples extends resource_js_1.APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "bulk", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new BulkAPI.Bulk(this._client)
        });
        Object.defineProperty(this, "validate", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ValidateAPI.Validate(this._client)
        });
    }
    /**
     * Create a new example.
     */
    create(body, options) {
        return this._client.post('/api/v1/examples', { body, ...options });
    }
    /**
     * Get a specific example.
     */
    retrieve(exampleID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/examples/${exampleID}`, { query, ...options });
    }
    /**
     * Update a specific example.
     */
    update(exampleID, body, options) {
        return this._client.patch((0, path_js_1.path) `/api/v1/examples/${exampleID}`, { body, ...options });
    }
    /**
     * Get all examples by query params
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/v1/examples', (pagination_js_1.OffsetPaginationTopLevelArray), {
            query,
            ...options,
        });
    }
    /**
     * Soft delete an example. Only deletes the example in the 'latest' version of the
     * dataset.
     */
    delete(exampleID, options) {
        return this._client.delete((0, path_js_1.path) `/api/v1/examples/${exampleID}`, options);
    }
    /**
     * Soft delete examples. Only deletes the examples in the 'latest' version of the
     * dataset.
     */
    deleteAll(params, options) {
        const { example_ids } = params;
        return this._client.delete('/api/v1/examples', { query: { example_ids }, ...options });
    }
    /**
     * Count all examples by query params
     */
    retrieveCount(query = {}, options) {
        return this._client.get('/api/v1/examples/count', { query, ...options });
    }
    /**
     * Upload examples from a CSV file.
     *
     * Note: For non-csv upload, please use the POST
     * /v1/platform/datasets/{dataset_id}/examples endpoint which provides more
     * efficient upload.
     */
    uploadFromCsv(datasetID, body, options) {
        return this._client.post((0, path_js_1.path) `/api/v1/examples/upload/${datasetID}`, (0, uploads_js_1.multipartFormRequestOptions)({ body, ...options }, this._client));
    }
}
exports.Examples = Examples;
Examples.Bulk = bulk_js_1.Bulk;
Examples.Validate = validate_js_1.Validate;
