// langsmith/dist/_openapi_client/resources/examples/examples.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import * as BulkAPI from './bulk.js';
import { Bulk } from './bulk.js';
import * as ValidateAPI from './validate.js';
import { Validate } from './validate.js';
import { OffsetPaginationTopLevelArray, } from '../../core/pagination.js';
import { multipartFormRequestOptions } from '../../internal/uploads.js';
import { path } from '../../internal/utils/path.js';
export class Examples extends APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "bulk", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new BulkAPI.Bulk(this._client)
        });
        Object.defineProperty(this, "validate", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ValidateAPI.Validate(this._client)
        });
    }
    /**
     * Create a new example.
     */
    create(body, options) {
        return this._client.post('/api/v1/examples', { body, ...options });
    }
    /**
     * Get a specific example.
     */
    retrieve(exampleID, query = {}, options) {
        return this._client.get(path `/api/v1/examples/${exampleID}`, { query, ...options });
    }
    /**
     * Update a specific example.
     */
    update(exampleID, body, options) {
        return this._client.patch(path `/api/v1/examples/${exampleID}`, { body, ...options });
    }
    /**
     * Get all examples by query params
     */
    list(query = {}, options) {
        return this._client.getAPIList('/api/v1/examples', (OffsetPaginationTopLevelArray), {
            query,
            ...options,
        });
    }
    /**
     * Soft delete an example. Only deletes the example in the 'latest' version of the
     * dataset.
     */
    delete(exampleID, options) {
        return this._client.delete(path `/api/v1/examples/${exampleID}`, options);
    }
    /**
     * Soft delete examples. Only deletes the examples in the 'latest' version of the
     * dataset.
     */
    deleteAll(params, options) {
        const { example_ids } = params;
        return this._client.delete('/api/v1/examples', { query: { example_ids }, ...options });
    }
    /**
     * Count all examples by query params
     */
    retrieveCount(query = {}, options) {
        return this._client.get('/api/v1/examples/count', { query, ...options });
    }
    /**
     * Upload examples from a CSV file.
     *
     * Note: For non-csv upload, please use the POST
     * /v1/platform/datasets/{dataset_id}/examples endpoint which provides more
     * efficient upload.
     */
    uploadFromCsv(datasetID, body, options) {
        return this._client.post(path `/api/v1/examples/upload/${datasetID}`, multipartFormRequestOptions({ body, ...options }, this._client));
    }
}
Examples.Bulk = Bulk;
Examples.Validate = Validate;
