// langsmith/dist/_openapi_client/resources/datasets/datasets.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import * as ComparativeAPI from './comparative.js';
import { Comparative, } from './comparative.js';
import * as ExperimentsAPI from './experiments.js';
import { Experiments } from './experiments.js';
import * as GroupAPI from './group.js';
import { Group } from './group.js';
import * as RunsAPI from './runs.js';
import { Runs, } from './runs.js';
import * as ShareAPI from './share.js';
import { Share } from './share.js';
import * as SplitsAPI from './splits.js';
import { Splits, } from './splits.js';
import * as VersionsAPI from './versions.js';
import { Versions, } from './versions.js';
import { OffsetPaginationTopLevelArray, } from '../../core/pagination.js';
import { multipartFormRequestOptions } from '../../internal/uploads.js';
import { path } from '../../internal/utils/path.js';
export class Datasets extends APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "versions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new VersionsAPI.Versions(this._client)
        });
        Object.defineProperty(this, "runs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RunsAPI.Runs(this._client)
        });
        Object.defineProperty(this, "group", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new GroupAPI.Group(this._client)
        });
        Object.defineProperty(this, "experiments", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ExperimentsAPI.Experiments(this._client)
        });
        Object.defineProperty(this, "share", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ShareAPI.Share(this._client)
        });
        Object.defineProperty(this, "comparative", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ComparativeAPI.Comparative(this._client)
        });
        Object.defineProperty(this, "splits", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new SplitsAPI.Splits(this._client)
        });
    }
    /**
     * Create a new dataset.
     */
    create(body, options) {
        return this._client.post('/api/v1/datasets', { body, ...options });
    }
    /**
     * Get a specific dataset.
     */
    retrieve(datasetID, options) {
        return this._client.get(path `/api/v1/datasets/${datasetID}`, options);
    }
    /**
     * Update a specific dataset.
     */
    update(datasetID, body, options) {
        return this._client.patch(path `/api/v1/datasets/${datasetID}`, { body, ...options });
    }
    /**
     * Get all datasets by query params and owner.
     */
    list(params = {}, options) {
        const { datatype, ...query } = params ?? {};
        return this._client.getAPIList('/api/v1/datasets', (OffsetPaginationTopLevelArray), {
            query: { data_type: datatype, ...query },
            ...options,
        });
    }
    /**
     * Delete a specific dataset.
     */
    delete(datasetID, options) {
        return this._client.delete(path `/api/v1/datasets/${datasetID}`, options);
    }
    /**
     * Clone a dataset.
     */
    clone(body, options) {
        return this._client.post('/api/v1/datasets/clone', { body, ...options });
    }
    /**
     * Download a dataset as CSV format.
     */
    retrieveCsv(datasetID, query = {}, options) {
        return this._client.get(path `/api/v1/datasets/${datasetID}/csv`, { query, ...options });
    }
    /**
     * Download a dataset as CSV format.
     */
    retrieveJSONL(datasetID, query = {}, options) {
        return this._client.get(path `/api/v1/datasets/${datasetID}/jsonl`, { query, ...options });
    }
    /**
     * Download a dataset as OpenAI Evals Jsonl format.
     */
    retrieveOpenAI(datasetID, query = {}, options) {
        return this._client.get(path `/api/v1/datasets/${datasetID}/openai`, { query, ...options });
    }
    /**
     * Download a dataset as OpenAI Jsonl format.
     */
    retrieveOpenAIFt(datasetID, query = {}, options) {
        return this._client.get(path `/api/v1/datasets/${datasetID}/openai_ft`, { query, ...options });
    }
    /**
     * Get dataset version by as_of or exact tag.
     */
    retrieveVersion(datasetID, query = {}, options) {
        return this._client.get(path `/api/v1/datasets/${datasetID}/version`, { query, ...options });
    }
    /**
     * Set a tag on a dataset version.
     */
    updateTags(datasetID, body, options) {
        return this._client.put(path `/api/v1/datasets/${datasetID}/tags`, { body, ...options });
    }
    /**
     * Create a new dataset from a CSV or JSONL file.
     */
    upload(body, options) {
        return this._client.post('/api/v1/datasets/upload', multipartFormRequestOptions({ body, ...options }, this._client));
    }
}
Datasets.Versions = Versions;
Datasets.Runs = Runs;
Datasets.Group = Group;
Datasets.Experiments = Experiments;
Datasets.Share = Share;
Datasets.Comparative = Comparative;
Datasets.Splits = Splits;
