// langsmith/dist/_openapi_client/resources/datasets/datasets.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Datasets = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const ComparativeAPI = __importStar(require("./comparative.cjs"));
const comparative_js_1 = require("./comparative.cjs");
const ExperimentsAPI = __importStar(require("./experiments.cjs"));
const experiments_js_1 = require("./experiments.cjs");
const GroupAPI = __importStar(require("./group.cjs"));
const group_js_1 = require("./group.cjs");
const RunsAPI = __importStar(require("./runs.cjs"));
const runs_js_1 = require("./runs.cjs");
const ShareAPI = __importStar(require("./share.cjs"));
const share_js_1 = require("./share.cjs");
const SplitsAPI = __importStar(require("./splits.cjs"));
const splits_js_1 = require("./splits.cjs");
const VersionsAPI = __importStar(require("./versions.cjs"));
const versions_js_1 = require("./versions.cjs");
const pagination_js_1 = require("../../core/pagination.cjs");
const uploads_js_1 = require("../../internal/uploads.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Datasets extends resource_js_1.APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "versions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new VersionsAPI.Versions(this._client)
        });
        Object.defineProperty(this, "runs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RunsAPI.Runs(this._client)
        });
        Object.defineProperty(this, "group", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new GroupAPI.Group(this._client)
        });
        Object.defineProperty(this, "experiments", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ExperimentsAPI.Experiments(this._client)
        });
        Object.defineProperty(this, "share", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ShareAPI.Share(this._client)
        });
        Object.defineProperty(this, "comparative", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new ComparativeAPI.Comparative(this._client)
        });
        Object.defineProperty(this, "splits", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new SplitsAPI.Splits(this._client)
        });
    }
    /**
     * Create a new dataset.
     */
    create(body, options) {
        return this._client.post('/api/v1/datasets', { body, ...options });
    }
    /**
     * Get a specific dataset.
     */
    retrieve(datasetID, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/datasets/${datasetID}`, options);
    }
    /**
     * Update a specific dataset.
     */
    update(datasetID, body, options) {
        return this._client.patch((0, path_js_1.path) `/api/v1/datasets/${datasetID}`, { body, ...options });
    }
    /**
     * Get all datasets by query params and owner.
     */
    list(params = {}, options) {
        const { datatype, ...query } = params ?? {};
        return this._client.getAPIList('/api/v1/datasets', (pagination_js_1.OffsetPaginationTopLevelArray), {
            query: { data_type: datatype, ...query },
            ...options,
        });
    }
    /**
     * Delete a specific dataset.
     */
    delete(datasetID, options) {
        return this._client.delete((0, path_js_1.path) `/api/v1/datasets/${datasetID}`, options);
    }
    /**
     * Clone a dataset.
     */
    clone(body, options) {
        return this._client.post('/api/v1/datasets/clone', { body, ...options });
    }
    /**
     * Download a dataset as CSV format.
     */
    retrieveCsv(datasetID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/datasets/${datasetID}/csv`, { query, ...options });
    }
    /**
     * Download a dataset as CSV format.
     */
    retrieveJSONL(datasetID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/datasets/${datasetID}/jsonl`, { query, ...options });
    }
    /**
     * Download a dataset as OpenAI Evals Jsonl format.
     */
    retrieveOpenAI(datasetID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/datasets/${datasetID}/openai`, { query, ...options });
    }
    /**
     * Download a dataset as OpenAI Jsonl format.
     */
    retrieveOpenAIFt(datasetID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/datasets/${datasetID}/openai_ft`, { query, ...options });
    }
    /**
     * Get dataset version by as_of or exact tag.
     */
    retrieveVersion(datasetID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/datasets/${datasetID}/version`, { query, ...options });
    }
    /**
     * Set a tag on a dataset version.
     */
    updateTags(datasetID, body, options) {
        return this._client.put((0, path_js_1.path) `/api/v1/datasets/${datasetID}/tags`, { body, ...options });
    }
    /**
     * Create a new dataset from a CSV or JSONL file.
     */
    upload(body, options) {
        return this._client.post('/api/v1/datasets/upload', (0, uploads_js_1.multipartFormRequestOptions)({ body, ...options }, this._client));
    }
}
exports.Datasets = Datasets;
Datasets.Versions = versions_js_1.Versions;
Datasets.Runs = runs_js_1.Runs;
Datasets.Group = group_js_1.Group;
Datasets.Experiments = experiments_js_1.Experiments;
Datasets.Share = share_js_1.Share;
Datasets.Comparative = comparative_js_1.Comparative;
Datasets.Splits = splits_js_1.Splits;
