// langsmith/dist/_openapi_client/resources/repos/repos.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Repos = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const DirectoriesAPI = __importStar(require("./directories.cjs"));
const directories_js_1 = require("./directories.cjs");
const pagination_js_1 = require("../../core/pagination.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Repos extends resource_js_1.APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "directories", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new DirectoriesAPI.Directories(this._client)
        });
    }
    /**
     * Create a repo.
     */
    create(body, options) {
        return this._client.post('/api/v1/repos', { body, ...options });
    }
    /**
     * Get a repo.
     */
    retrieve(repo, params, options) {
        const { owner } = params;
        return this._client.get((0, path_js_1.path) `/api/v1/repos/${owner}/${repo}`, options);
    }
    /**
     * Update a repo.
     */
    update(repo, params, options) {
        const { owner, ...body } = params;
        return this._client.patch((0, path_js_1.path) `/api/v1/repos/${owner}/${repo}`, { body, ...options });
    }
    /**
     * Get all repos.
     */
    list(params = {}, options) {
        const { single_repo_type, ...query } = params ?? {};
        return this._client.getAPIList('/api/v1/repos', (pagination_js_1.OffsetPaginationRepos), {
            query: { repo_type: single_repo_type, ...query },
            ...options,
        });
    }
    /**
     * Delete a repo.
     */
    delete(repo, params, options) {
        const { owner } = params;
        return this._client.delete((0, path_js_1.path) `/api/v1/repos/${owner}/${repo}`, options);
    }
}
exports.Repos = Repos;
Repos.Directories = directories_js_1.Directories;
