// langsmith/dist/_openapi_client/resources/repos/repos.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import * as DirectoriesAPI from './directories.js';
import { Directories, } from './directories.js';
import { OffsetPaginationRepos } from '../../core/pagination.js';
import { path } from '../../internal/utils/path.js';
export class Repos extends APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "directories", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new DirectoriesAPI.Directories(this._client)
        });
    }
    /**
     * Create a repo.
     */
    create(body, options) {
        return this._client.post('/api/v1/repos', { body, ...options });
    }
    /**
     * Get a repo.
     */
    retrieve(repo, params, options) {
        const { owner } = params;
        return this._client.get(path `/api/v1/repos/${owner}/${repo}`, options);
    }
    /**
     * Update a repo.
     */
    update(repo, params, options) {
        const { owner, ...body } = params;
        return this._client.patch(path `/api/v1/repos/${owner}/${repo}`, { body, ...options });
    }
    /**
     * Get all repos.
     */
    list(params = {}, options) {
        const { single_repo_type, ...query } = params ?? {};
        return this._client.getAPIList('/api/v1/repos', (OffsetPaginationRepos), {
            query: { repo_type: single_repo_type, ...query },
            ...options,
        });
    }
    /**
     * Delete a repo.
     */
    delete(repo, params, options) {
        const { owner } = params;
        return this._client.delete(path `/api/v1/repos/${owner}/${repo}`, options);
    }
}
Repos.Directories = Directories;
