// langsmith/dist/sandbox/mounts.js
import { proxyConfig } from "./proxy_config.js";
function requireNonEmptyString(value, field) {
    if (typeof value !== "string" || value.trim() === "") {
        throw new Error(`${field} must be a non-empty string`);
    }
    return value.trim();
}
export function s3Mount({ id, mountPath, bucket, region = "us-east-1", prefix, endpointUrl = "https://s3.amazonaws.com", pathStyle = false, readOnly, cache, }) {
    const mount = {
        id: requireNonEmptyString(id, "id"),
        type: "s3",
        mount_path: requireNonEmptyString(mountPath, "mountPath"),
        s3: {
            endpoint_url: requireNonEmptyString(endpointUrl, "endpointUrl"),
            region: requireNonEmptyString(region, "region"),
            bucket: requireNonEmptyString(bucket, "bucket"),
            path_style: pathStyle,
        },
    };
    if (prefix !== undefined) {
        mount.s3.prefix = requireNonEmptyString(prefix, "prefix");
    }
    if (readOnly !== undefined) {
        mount.read_only = readOnly;
    }
    if (cache !== undefined) {
        mount.cache = { ...cache };
    }
    return mount;
}
export function gcsMount({ id, mountPath, bucket, prefix, readOnly, cache, }) {
    const mount = {
        id: requireNonEmptyString(id, "id"),
        type: "gcs",
        mount_path: requireNonEmptyString(mountPath, "mountPath"),
        gcs: {
            bucket: requireNonEmptyString(bucket, "bucket"),
        },
    };
    if (prefix !== undefined) {
        mount.gcs.prefix = requireNonEmptyString(prefix, "prefix");
    }
    if (readOnly !== undefined) {
        mount.read_only = readOnly;
    }
    if (cache !== undefined) {
        mount.cache = { ...cache };
    }
    return mount;
}
function normalizeMounts(mounts) {
    if (!Array.isArray(mounts) || mounts.length === 0) {
        throw new Error("mounts must be a non-empty array of mount objects");
    }
    return mounts.map((mount) => {
        if (mount === null || typeof mount !== "object" || Array.isArray(mount)) {
            throw new Error("mounts must be a non-empty array of mount objects");
        }
        if (mount.type !== "s3" && mount.type !== "gcs") {
            throw new Error("mountConfig only supports s3 and gcs mounts");
        }
        return mount;
    });
}
function normalizeAuthRules(auth) {
    if (!Array.isArray(auth)) {
        throw new Error("auth must be an array of provider auth rules");
    }
    const byProvider = {};
    for (const rule of auth) {
        if (rule === null || typeof rule !== "object" || Array.isArray(rule)) {
            throw new Error("auth must be an array of provider auth rules");
        }
        const provider = rule.type;
        if (provider !== "aws" && provider !== "gcp") {
            throw new Error("mountConfig auth only supports aws and gcp rules");
        }
        if (byProvider[provider] !== undefined) {
            throw new Error(`duplicate ${provider} auth rule in mountConfig`);
        }
        byProvider[provider] = rule;
    }
    return byProvider;
}
export function mountConfig({ auth, mounts, }) {
    const normalizedMounts = normalizeMounts(mounts);
    const authByProvider = normalizeAuthRules(auth);
    const mountProviders = new Set(normalizedMounts.map((mount) => mount.type));
    if (mountProviders.has("s3") && authByProvider.aws === undefined) {
        throw new Error("s3 mounts require aws auth in mountConfig");
    }
    if (mountProviders.has("gcs") && authByProvider.gcp === undefined) {
        throw new Error("gcs mounts require gcp auth in mountConfig");
    }
    if (authByProvider.aws !== undefined && !mountProviders.has("s3")) {
        throw new Error("aws auth requires at least one s3 mount in mountConfig");
    }
    if (authByProvider.gcp !== undefined && !mountProviders.has("gcs")) {
        throw new Error("gcp auth requires at least one gcs mount in mountConfig");
    }
    return {
        mounts: normalizedMounts,
        proxyConfig: proxyConfig({
            rules: ["aws", "gcp"]
                .map((provider) => authByProvider[provider])
                .filter((rule) => rule !== undefined),
        }),
    };
}
