// langsmith/dist/_openapi_client/resources/runs/runs.cjs
"use strict";
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runs = void 0;
const resource_js_1 = require("../../core/resource.cjs");
const RulesAPI = __importStar(require("./rules.cjs"));
const rules_js_1 = require("./rules.cjs");
const path_js_1 = require("../../internal/utils/path.cjs");
class Runs extends resource_js_1.APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "rules", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RulesAPI.Rules(this._client)
        });
    }
    /**
     * Queues a single run for ingestion. The request body must be a JSON-encoded run
     * object that follows the Run schema.
     */
    create(body, options) {
        return this._client.post('/runs', { body, ...options });
    }
    /**
     * Get a specific run.
     */
    retrieve(runID, query = {}, options) {
        return this._client.get((0, path_js_1.path) `/api/v1/runs/${runID}`, { query, ...options });
    }
    /**
     * Updates a run identified by its ID. The body should contain only the fields to
     * be changed; unknown fields are ignored.
     */
    update(runID, body, options) {
        return this._client.patch((0, path_js_1.path) `/runs/${runID}`, { body, ...options });
    }
    /**
     * Ingests a batch of runs in a single JSON payload. The payload must have `post`
     * and/or `patch` arrays containing run objects. Prefer this endpoint over
     * single‑run ingestion when submitting hundreds of runs, but `/runs/multipart`
     * offers better handling for very large fields and attachments.
     */
    ingestBatch(body, options) {
        return this._client.post('/runs/batch', { body, ...options });
    }
    /**
     * Query Runs
     */
    query(body, options) {
        return this._client.post('/api/v1/runs/query', { body, ...options });
    }
    /**
     * Get all runs by query in body payload.
     */
    stats(body, options) {
        return this._client.post('/api/v1/runs/stats', { body, ...options });
    }
    /**
     * Update a run.
     */
    update2(runID, options) {
        return this._client.patch((0, path_js_1.path) `/api/v1/runs/${runID}`, options);
    }
}
exports.Runs = Runs;
Runs.Rules = rules_js_1.Rules;
