// langsmith/dist/_openapi_client/resources/runs/runs.js
// @ts-nocheck
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from '../../core/resource.js';
import * as RulesAPI from './rules.js';
import { Rules } from './rules.js';
import { path } from '../../internal/utils/path.js';
export class Runs extends APIResource {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "rules", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new RulesAPI.Rules(this._client)
        });
    }
    /**
     * Queues a single run for ingestion. The request body must be a JSON-encoded run
     * object that follows the Run schema.
     */
    create(body, options) {
        return this._client.post('/runs', { body, ...options });
    }
    /**
     * Get a specific run.
     */
    retrieve(runID, query = {}, options) {
        return this._client.get(path `/api/v1/runs/${runID}`, { query, ...options });
    }
    /**
     * Updates a run identified by its ID. The body should contain only the fields to
     * be changed; unknown fields are ignored.
     */
    update(runID, body, options) {
        return this._client.patch(path `/runs/${runID}`, { body, ...options });
    }
    /**
     * Ingests a batch of runs in a single JSON payload. The payload must have `post`
     * and/or `patch` arrays containing run objects. Prefer this endpoint over
     * single‑run ingestion when submitting hundreds of runs, but `/runs/multipart`
     * offers better handling for very large fields and attachments.
     */
    ingestBatch(body, options) {
        return this._client.post('/runs/batch', { body, ...options });
    }
    /**
     * Query Runs
     */
    query(body, options) {
        return this._client.post('/api/v1/runs/query', { body, ...options });
    }
    /**
     * Get all runs by query in body payload.
     */
    stats(body, options) {
        return this._client.post('/api/v1/runs/stats', { body, ...options });
    }
    /**
     * Update a run.
     */
    update2(runID, options) {
        return this._client.patch(path `/api/v1/runs/${runID}`, options);
    }
}
Runs.Rules = Rules;
