// @langchain/google-genai/dist/utils/stream_events.cjs
const require_common = require("./common.cjs");
let _langchain_core_language_models_compat = require("@langchain/core/language_models/compat");
//#region src/utils/stream_events.ts
async function* convertGoogleGenAIStream(source, options = {}) {
	const shouldStreamUsage = options.streamUsage ?? true;
	const blockAccumulators = /* @__PURE__ */ new Map();
	const blockKeyToIndex = /* @__PURE__ */ new Map();
	let nextBlockIndex = 0;
	let messageStarted = false;
	let usageSnapshot;
	let finishReason = "stop";
	const getOrCreateBlockIndex = (key, initial) => {
		const existing = blockKeyToIndex.get(key);
		if (existing !== void 0) return {
			index: existing,
			isNew: false
		};
		const index = nextBlockIndex++;
		blockKeyToIndex.set(key, index);
		blockAccumulators.set(index, { ...initial });
		return {
			index,
			isNew: true
		};
	};
	for await (const response of source) {
		if (!messageStarted) {
			messageStarted = true;
			yield { event: "message-start" };
		}
		if (shouldStreamUsage && "usageMetadata" in response && response.usageMetadata) {
			usageSnapshot = require_common.convertUsageMetadata(response.usageMetadata, options.model ?? "");
			yield {
				event: "usage",
				usage: usageSnapshot
			};
		}
		const candidate = response.candidates?.[0];
		if (candidate?.finishReason) finishReason = mapGeminiFinishReason(candidate.finishReason);
		const parts = candidate?.content?.parts;
		if (!parts) continue;
		let toolIdx = 0;
		for (const part of parts) if ("text" in part && part.text) if ("thought" in part && part.thought) {
			const { index, isNew } = getOrCreateBlockIndex("reasoning", {
				type: "reasoning",
				reasoning: ""
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "reasoning",
					reasoning: ""
				}
			};
			const acc = blockAccumulators.get(index);
			acc.reasoning = (acc.reasoning ?? "") + part.text;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "reasoning-delta",
					reasoning: part.text
				}
			};
		} else {
			const { index, isNew } = getOrCreateBlockIndex("text", {
				type: "text",
				text: ""
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "text",
					text: ""
				}
			};
			const acc = blockAccumulators.get(index);
			acc.text = (acc.text ?? "") + part.text;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "text-delta",
					text: part.text
				}
			};
		}
		else if ("functionCall" in part && part.functionCall) {
			const key = `tool:${toolIdx}`;
			const args = JSON.stringify(part.functionCall.args ?? {});
			const { index, isNew } = getOrCreateBlockIndex(key, {
				type: "tool_call_chunk",
				name: part.functionCall.name,
				args: "",
				index: toolIdx
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "tool_call_chunk",
					name: part.functionCall.name,
					args: "",
					index: toolIdx
				}
			};
			const acc = blockAccumulators.get(index);
			acc.args = args;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "block-delta",
					fields: {
						type: "tool_call_chunk",
						name: acc.name,
						args: acc.args
					}
				}
			};
			toolIdx += 1;
		}
	}
	for (const [index, acc] of blockAccumulators) yield {
		event: "content-block-finish",
		index,
		content: (0, _langchain_core_language_models_compat.finalizeContentBlock)(acc)
	};
	yield {
		event: "message-finish",
		reason: finishReason,
		...usageSnapshot ? { usage: usageSnapshot } : {},
		responseMetadata: { model_provider: "google-genai" }
	};
}
function mapGeminiFinishReason(reason) {
	switch (reason.toLowerCase()) {
		case "max_tokens":
		case "max-token":
		case "max_token": return "length";
		case "safety":
		case "recitation":
		case "language":
		case "blocklist":
		case "prohibited_content":
		case "prohibited-content":
		case "spii":
		case "image_safety":
		case "image-safety":
		case "image_prohibited_content":
		case "image-prohibited-content":
		case "image_recitation":
		case "image-recitation": return "content_filter";
		default: return "stop";
	}
}
//#endregion
exports.convertGoogleGenAIStream = convertGoogleGenAIStream;
